File: 393-utf-8-validation/ken/393.py
class Solution(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """

        valid = True
        i = 0

        while i < len(data) and valid:
            binText = format(data[i], 'b').zfill(8)
            i += 1
            num = 0

            if binText.startswith("0"):
                continue
            elif binText.startswith("110"):
                num = 1
            elif binText.startswith("1110"):
                num = 2
            elif binText.startswith("11110"):
                num = 3
            else:
                valid = False
                break

            if len(data) - i < num:
                valid = False
                break

            for j in xrange(num):
                binText = format(data[i], 'b').zfill(8)
                i += 1

                if not binText.startswith("10"):
                    valid = False
                    break

        return valid

File: 393-utf-8-validation/leochang/393.cpp
class Solution {
public:
	bool validUtf8(vector<int>& data) {
		int byteCounts = 0;
		for (int i = 0; i < data.size(); i++) {
			if (data[i] < 128) {
				if (byteCounts > 0) return false;
				continue;
			}
			if (data[i] >= 128 && data[i] < 192) {
				byteCounts--;
				continue;
			}
			if (byteCounts > 0 || data[i] > 247) return false;
			if (data[i] >= 240) {
				byteCounts+=3;
				continue;
			}
			if (data[i] >= 224) {
				byteCounts+=2;
				continue;
			}
			if (data[i] >= 192) {
				byteCounts+=1;
				continue;
			}
		}
		return 0 == byteCounts;
	}
};
